File: 0077-combinations/0077-combinations.cpp
class Solution {
public:
    void recur_combine(int n, int k, int cur, vector<int>& sofar, vector<vector<int>>& ans) {
    if (sofar.size() == k) {
        ans.push_back(sofar);
        return;
    }

    for (int i = cur; i <= n; i++) {
        sofar.push_back(i);

        recur_combine(n, k, i+1, sofar, ans);

        sofar.pop_back();
    }
}

vector<vector<int>> combine(int n, int k) {
    vector<int> sofar;
    vector<vector<int>> ans;

    recur_combine(n, k, 1, sofar, ans);
    return ans;
}
};

File: 0077-combinations/0077-combinations.cs
public class Solution {
    public IList<IList<int>> Combine(int n, int k)
{
    IList<IList<int>> ans = new List<IList<int>>();
    IList<int> sofar = new List<int>();

    recur_combine(n, k, 1, sofar, ans);
    return ans;
}

public void recur_combine(int n, int k, int cur, IList<int> sofar, IList<IList<int>> ans)
{
    if (sofar.Count == k)
    {
        ans.Add(new List<int>(sofar));
        return;
    }

    for (int i = cur; i <= n; i++)
    {
        sofar.Add(i);

        recur_combine(n, k, i + 1, sofar, ans);

        if (sofar.Any())
            sofar.RemoveAt(sofar.Count - 1);
    }
}
}